Lesson 7 ajax part2

ignatbagaev opened the pull request.

ignatbagaev committed:
user could update own question

ignatbagaev committed:
user could update own answer

ignatbagaev committed:
fix questions#update

Fixed questions#update action to stay on :edit view if update was aborted.

ignatbagaev committed:
delete answer via ajax

ignatbagaev committed:
mark answer as accepted

User could mark one answer belongs to his question as accepted

ignatbagaev committed:
accept another answer

ignatbagaev committed:
update answer using ajax

ignatbagaev committed:
update question using ajax and replace selenium webdriver with capybara webkit

ignatbagaev committed:
add rubocop and make some code refactoring

ignatbagaev committed:
refactor answer accepting

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -1,16 +1,29 @@
+  def accept
+    question = @answer.question
+    if current_user.author_of? question
+      question.answers.update_all(status: 0)
```
Эта логика должна быть в методе модели Answer, а тут только вызывается этот метод. И статус надо ставить  не 1/0, а true/false.  И еще эту логику надо в транзакцию обернуть, чтобы изменения были атомарными

vkurennov commented on app/models/answer.rb:
```diff
@@ -1,4 +1,8 @@
 class Answer < ActiveRecord::Base
+  default_scope { order(status: :desc) }
+
+  enum status: { common: 0, accepted: 1 }
```
Зачем тут enum, когда достатончо обычного флага типа boolean?

vkurennov commented on app/views/answers/_answer.html.slim:
```diff
@@ -1,8 +1,23 @@
+      h5= answer.body
+      h6.text-right
+        - if (current_user && current_user.author_of?(@question)) && answer.accepted?
+          = button_to "Accept", accept_answer_path(answer), remote: true, class: 'btn btn-default button-accept accept-hidden', id: "accept-#{answer.id}"
```
Пожалуй, текст ссылки тут должен быть другой

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -1,16 +1,29 @@
-    @answer = Answer.find(params[:id])
-    destroy_answer
-    redirect_to @answer.question
+    @answer.destroy if current_user.author_of?(@answer)
```
А если не автор? destory выполнен не будет, а вид все равно отрендерится, т.к. если не указано иное, контроллер рендерит одноименный с экшеном вид.

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -1,16 +1,29 @@
   end
 
+  def update
+    if current_user.author_of? @answer
```
А если это не автор, что тогда?

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -1,16 +1,29 @@
+
+  def accept
+    question = @answer.question
+    if current_user.author_of? question
```
А если это не автор вопроса, что тогда?

vkurennov commented on app/controllers/questions_controller.rb:
```diff
@@ -15,12 +15,15 @@ def index
+  end
+
+  def update
+    @question.update(question_params) if current_user.author_of? @question
```
А если это не автор, что тогда?

vkurennov commented on app/views/answers/update.js.erb:
```diff
@@ -0,0 +1,14 @@
+    $('div#answer-error-' + id).html('<%= j message %>');
+  <% end %>
+<% else %>
+  $('.answers').html('<%= j render 'questions/answers' %>');
```
Нужно обновлять только один ответ,  а не перерисовыывать весь список.

vkurennov commented on app/views/answers/update.js.erb:
```diff
@@ -0,0 +1,14 @@
+  <% end %>
+<% else %>
+  $('.answers').html('<%= j render 'questions/answers' %>');
+  $('.edit-answer-link').click(function(e) {
```
Прочитай [вот это](http://connect.thinknetica.com/t/ajax-turbolinks/955) и избавься от дублирования в js-ответах

vkurennov commented on app/views/questions/update.js.erb:
```diff
@@ -0,0 +1,13 @@
+  <% end %>
+<% else %>
+  $('.question-item').html( '<%= j render 'question' %>');
+  $('.edit-question-link').click(function(e) {
```
Прочитай [вот это](http://connect.thinknetica.com/t/ajax-turbolinks/955) и избавься от дублирования в js-ответах

vkurennov commented on spec/acceptance/answer_specs/edit_answer_spec.rb:
```diff
@@ -0,0 +1,46 @@
+  before { question.answers << answer }
+  before(:each) { log_in user }
+  
+  scenario 'link to edit', js: true do
```
Этот сценарий лишний, т.к. его поведение неявно проверяется в следующем. Зато здесь должен быть сценарий, проверяющий, что не-автор не видит ссылки на редактирование.

vkurennov commented on spec/acceptance/question_specs/edit_question.spec.rb:
```diff
@@ -0,0 +1,43 @@
+    click_link "Edit"
+  end
+  
+  scenario 'link to edit', js: true do
```
Этот сценарий лишний, т.к. его поведение неявно проверяется в следующем. Зато здесь должен быть сценарий, проверяющий, что не-автор не видит ссылки на редактирование.

vkurennov commented:
Исправь замечания

ignatbagaev committed:
change answers status:integer attribute to accepted:boolean

ignatbagaev committed:
move business logic of answer accepting from controller to model

ignatbagaev committed:
fixes according to comments on PR

Fixed rendering after failing of action(answers#update, answers#destroy, answers#accept and questions#update). Also made some code refactoring in views, acceptance specs and in assets

ignatbagaev committed:
fix answers#accept

ignatbagaev committed:
small refactoring in answers controller to make it more DRY

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -19,11 +31,15 @@ def answer_params
+    if current_user.author_of? object
+      yield if block_given?
+    else
+      render nothing: true, status: 401
```
Это можно короче записать: `head 401`, и тут лучше подойдет 403 статус (forbidden), а не 401, т.к. 401 используется, если юзер не залогинен, а 403 - если залогинен, но не имеет прав на выполнение действия

vkurennov commented on app/models/question.rb:
```diff
@@ -1,6 +1,10 @@
   validates :title, :body, :user_id, presence: true
+
+  def has_accepted_answer?
+    answers.pluck(:accepted).include? true
```
Не очень эффективно. Метод pluck выбирает данные из базы, а можно это сделать без получания самих данных, т.к. они все равно не нужны: `answers.where(accepted: true).exists?`

vkurennov commented on app/views/answers/_answer.html.slim:
```diff
@@ -1,8 +1,23 @@
+      h5= answer.body
+      h6.text-right
+        - if (current_user && current_user.author_of?(answer.question)) && answer.accepted?
+          = button_to "Best", accept_answer_path(answer), remote: true, class: 'btn btn-default button-accept accept-hidden', id: "accept-#{answer.id}"
```
Почему у тебя и для принятого и для не принятого ответа кнопка называется одинаково - "Best"? Наверное, нужно их по-разному назвать, в зависимости от состояния ответа

vkurennov commented:
Принято, но поправь замечания перед выполнением следующего задания

ignatbagaev committed:
small fixes with accpt button and refactoring
